application.go

```diff
@@ -0,0 +1,144 @@
+// For more information please visit https://mesg.com.
+package mesg
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"log"
+	"os"
+	"time"
+
+	"github.com/mesg-foundation/core/api/core"
+	"google.golang.org/grpc"
+)
+
+const (
+	endpointEnv     = "MESG_ENDPOINT"
+	defaultEndpoint = "localhost:50052"
+)
+
+// Application represents is a MESG application.
+type Application struct {
+	endpoint string
+
+	// Client is the gRPC core client of MESG.
+	client core.CoreClient
+	conn   *grpc.ClientConn
+
+	callTimeout time.Duration
+
+	dialOptions []grpc.DialOption
+
+	log       *log.Logger
+	logOutput io.Writer
+}
+
+// Option is the configuration func of Application.
+type Option func(*Application)
+
+// New returns a new Application with options.
+func New(options ...Option) (*Application, error) {
+	a := &Application{
+		endpoint:    os.Getenv(endpointEnv),
+		callTimeout: time.Second * 10,
+		logOutput:   os.Stdout,
+		dialOptions: []grpc.DialOption{grpc.WithInsecure()},
+	}
+	for _, option := range options {
+		option(a)
+	}
+	a.log = log.New(a.logOutput, "mesg", log.LstdFlags)
+	if a.endpoint == "" {
+		a.endpoint = defaultEndpoint
+	}
+	return a, a.setupCoreClient()
+}
+
+// EndpointOption receives the endpoint of MESG.
+func EndpointOption(address string) Option {
+	return func(a *Application) {
+		a.endpoint = address
+	}
+}
+
+// LogOutputOption uses out as a log destination.
+func LogOutputOption(out io.Writer) Option {
+	return func(app *Application) {
+		app.logOutput = out
+	}
+}
+
+// DialOption used to mock socket communication for unit testing.
+func DialOption(dialer Dialer) Option {
+	return func(a *Application) {
+		a.dialOptions = append(a.dialOptions, grpc.WithDialer(newGRPCDialer(dialer).Dial))
+	}
+}
+
+func (a *Application) setupCoreClient() error {
+	var err error
+	ctx, cancel := context.WithTimeout(context.Background(), a.callTimeout)
+	defer cancel()
+	a.conn, err = grpc.DialContext(ctx, a.endpoint, a.dialOptions...)
+	if err != nil {
+		return err
+	}
+	a.client = core.NewCoreClient(a.conn)
+	return nil
+}
+
+// execute executes a task for serviceID with given data.
+func (a *Application) execute(serviceID, task string, data Data) (executionID string, err error) {
+	inputDataBytes, err := json.Marshal(data)
+	if err != nil {
+		return "", err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), a.callTimeout)
+	defer cancel()
+	resp, err := a.client.ExecuteTask(ctx, &core.ExecuteTaskRequest{
+		ServiceID: serviceID,
+		TaskKey:   task,
+		InputData: string(inputDataBytes),
+	})
+	if err != nil {
+		return "", err
+	}
+	return resp.ExecutionID, nil
+}
+
+// startServices starts mesg services.
+func (a *Application) startServices(ids ...string) error {
+	idsLen := len(ids)
+	if idsLen == 0 {
+		return nil
+	}
+
+	errC := make(chan error, 0)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for _, id := range ids {
+		go a.startService(ctx, id, errC)
```

Comment:
This is totally valid, channel returned from `Context.Done()` will be closed right after `cancel()` called or timeout reached. `dockerClient.StartService()` internally waits a signal from done channel to abandon its execution and stop blocking. Since a closed channel always emits zero values to its readers, every listener of done channel will get a zero value immediately. In this case all `dockerClient.StartService()` calls will receive a signal from done channel and stop blocking by returning a canceled or deadline exceeded error.

About the implementation of `Application.startServices()`:

* N service start requests will be made to start N services concurrently. If timeout exceeds, all continuing start requests will be abandoned inside of `dockerClient.StartService()` because done channel will receive a signal and code will return with a deadline exceeded error.
* After each return from `dockerClient.StartService()` calls, `errC` will receive a nil/deadline exceed/canceled or an API related error. Reader of `errC`, checks for a non nil error and `Application.startServices()` immediately returns with it.
* After `Application.startServices()` returned, deferred `cancel()` called to ensure to abandon all continuing start task requests. Of course, if the returned error was a deadline exceed error we could safely assume that all start task requests will be canceled. But if timeout is not exceed and `Application.startServices()` completed due to an API error got from `errC`, we have to abandon all running task requests by calling `cancel()`.

Btw, after the second check, I see that we have a goroutine leak in the sender side of `errC` because we early return after a first non nil error received from `errC` and all senders blocks forever. I'll update `errC`'s buffer size with `idsLen` so all senders can send their error values to channel without blocking and `errC` can get garbage collected after `Application.startServices()` returned.